pkg/cache/lru/lru.go

```diff
@@ -113,10 +113,10 @@ func (c *LRU[K, V]) removeOldest() {
 func (c *LRU[K, V]) removeElement(e *list.Element) {
 	c.evictList.Remove(e)
 	kv := e.Value.(entry[K, V])
-	delete(c.items, kv.key)
 	if c.onEvicted != nil {
 		c.onEvicted(kv.key, kv.value)
 	}
+	delete(c.items, kv.key)
 	c.metrics.evictions.Inc()
 }
```

Comment:
Related to the comment above, the `Purge()` method could be incorrect, though, as no reference kept and the `onEvicted` and `delete` calls will run concurrently with the code in `main`

Reply:
@kakkoyun realised that this is not correct, we should have a reference to key and value